torchgeo/datasets/rioxarray.py

```diff
@@ -0,0 +1,156 @@
+
+                    (minx, miny, maxx, maxy) = ds.rio.bounds()
+
+                if hasattr(ds, "time"):
+                    try:
+                        indices = ds.indexes["time"].to_datetimeindex()
+                    except AttributeError:
+                        indices = ds.indexes["time"]
+
+                    mint = indices.min().to_pydatetime().timestamp()
+                    maxt = indices.max().to_pydatetime().timestamp()
+                else:
+                    mint = 0
+                    maxt = sys.maxsize
+                coords = (minx, maxx, miny, maxy, mint, maxt)
+                self.index.insert(i, coords, filepath)
+                i += 1
+
+        if i == 0:
+            msg = f"No {self.__class__.__name__} data was found in `root='{self.root}'`"
+            raise FileNotFoundError(msg)
+
+        if not crs:
+            self._crs = "EPSG:4326"
+        else:
+            self._crs = cast(CRS, crs)
+        self.res = cast(float, res)
+
+    def __getitem__(self, query: BoundingBox) -> dict[str, Any]:
+        """Retrieve image/mask and metadata indexed by query.
+
+        Args:
+            query: (minx, maxx, miny, maxy, mint, maxt) coordinates to index
+
+        Returns:
+            sample of image/mask and metadata at that index
+
+        Raises:
+            IndexError: if query is not found in the index
+        """
+        hits = self.index.intersection(tuple(query), objects=True)
+        items = [hit.object for hit in hits]
+
+        if not items:
+            raise IndexError(
+                f"query: {query} not found in index with bounds: {self.bounds}"
+            )
+
+        data_arrays: list["np.typing.NDArray[np.float32]"] = []
+        for item in items:
+            with xr.open_dataset(item, decode_cf=True) as ds:
+                if not ds.rio.crs:
+                    ds.rio.write_crs(self._crs, inplace=True)
+
+                # clip box ignores time dimension
+                clipped = ds.rio.clip_box(
+                    minx=query.minx, miny=query.miny, maxx=query.maxx, maxy=query.maxy
+                )
+                # select time dimension
+                if hasattr(ds, "time"):
+                    try:
+                        clipped["time"] = clipped.indexes["time"].to_datetimeindex()
+                    except AttributeError:
+                        clipped["time"] = clipped.indexes["time"]
+                    clipped = clipped.sel(
+                        time=slice(
+                            datetime.fromtimestamp(query.mint),
+                            datetime.fromtimestamp(query.maxt),
+                        )
+                    )
+                for variable in self.data_variables:
+                    if hasattr(clipped, variable):
+                        data_arrays.append(clipped[variable].data.squeeze())
+
+        sample = {"image": torch.from_numpy(np.stack(data_arrays)), "bbox": query}
```

Comment:
We should support both images and masks. See the `is_image` attribute in `RasterDataset` for how we do this there. You can also copy the `dtype` property to automatically choose what dtype to cast to.

Reply:
I'm not convinced this works correctly for multiple overlapping files. We shouldn't be stacking, we should be merging.

Reply:
You are right, maybe have to rethink the base class thing and have one for xarray climate data that is not using crs explicitly (`class XarrayDataset(GeoDataset)`)  and one that is intended for crs depending data sources like MODIS and more similar to `RasterDataset` and using `rioxarray` so the current naming `class RioXarrayDataset(GeoDataset)` but with the required functionality to handle overlapping files etc.